tests/src/com.github.ben-manes.caffeine/caffeine/2.9.3/src/test/java/com_github_ben_manes_caffeine/caffeine/CaffeineTest.java

```diff
@@ -0,0 +1,101 @@
+/*
+ * Copyright and related rights waived via CC0
+ *
+ * You should have received a copy of the CC0 legalcode along with this
+ * work. If not, see <http://creativecommons.org/publicdomain/zero/1.0/>.
+ */
+package com_github_ben_manes_caffeine.caffeine;
+
+import com.github.benmanes.caffeine.cache.Cache;
+import com.github.benmanes.caffeine.cache.Caffeine;
+import com.github.benmanes.caffeine.cache.CaffeineSpec;
+import com.github.benmanes.caffeine.cache.LoadingCache;
+import com.github.benmanes.caffeine.cache.Scheduler;
+import com.github.benmanes.caffeine.cache.stats.CacheStats;
+import com.google.common.testing.FakeTicker;
+import org.junit.jupiter.api.Test;
+
+import java.lang.ref.Cleaner;
+import java.time.Duration;
+import java.util.concurrent.CompletableFuture;
+import java.util.concurrent.TimeUnit;
+
+import static org.assertj.core.api.Assertions.assertThat;
+import static org.awaitility.Awaitility.await;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNotEquals;
+
+public class CaffeineTest {
+    @Test
+    void testRefresh() {
+        TestRefreshLoader testRefreshLoader = new TestRefreshLoader();
+        LoadingCache<String, CompletableFuture<String>> cache = Caffeine.newBuilder()
+                .refreshAfterWrite(Duration.ofMillis(100))
+                .build(testRefreshLoader::load);
+        cache.get("Hello");
+        CompletableFuture<String> firstValue = cache.getIfPresent("Hello");
+        assertThat(firstValue).isNotNull();
+        assertEquals("World", firstValue.join());
+        testRefreshLoader.setFlag(true);
+        cache.refresh("Hello");
+        await().atMost(Duration.ofSeconds(5)).untilAsserted(() -> {
+            CompletableFuture<String> secondValue = cache.getIfPresent("Hello");
+            assertThat(secondValue).isNotNull();
+            assertEquals("Universe", secondValue.join());
+        });
+    }
+
+    @Test
+    void testStatistics() {
+        Cache<String, String> graphs = Caffeine.newBuilder().maximumSize(10_000).recordStats().build();
+        graphs.put("Hello", "World");
+        assertThat(graphs.getIfPresent("Hello")).isEqualTo("World");
+        CacheStats cacheStats = graphs.stats();
+        assertThat(cacheStats.hitCount()).isEqualTo(1);
+        assertThat(cacheStats.missCount()).isEqualTo(0);
+        assertThat(cacheStats.evictionCount()).isEqualTo(0);
+    }
+
+    @Test
+    void testSpecification() {
+        CaffeineSpec spec = CaffeineSpec.parse("maximumWeight=1000, expireAfterWrite=10m, recordStats");
```

Comment:
Is `recordStats` a leftover here? Again, it looks like no stats are used in this test. Since you exercised stats in previous test, I guess we can remove it?

Reply:
- I remembered an issue involved in this unit test based on the comments I wrote for the related unit test a year ago. NativeTest fails when using `CaffeineSpec spec = CaffeineSpec.parse("maximumWeight=1000, expireAfterWrite=10m");`. But NativeTest works fine when using `CaffeineSpec spec = CaffeineSpec.parse("maximumWeight=1000, expireAfterWrite=10m, recordStats");`.
- I have now revisited the source code of `com.github.benmanes.caffeine.cache.CaffeineSpec` and I realized that there is a missing JSON entry. This is not automatically collected by Tracing Agent.
```json
{
   "condition":{"typeReachable":"com.github.benmanes.caffeine.cache.NodeFactory"},
   "name":"com.github.benmanes.caffeine.cache.SSMWW",
   "methods":[{"name":"<init>","parameterTypes":["com.github.benmanes.caffeine.cache.Caffeine","com.github.benmanes.caffeine.cache.CacheLoader"," boolean"] }]
}
```
- I removed `CaffeineSpec`'s use of `recordStats` and added the relevant JSON.